### PR TITLE
support str|None , mapped_column, AnyURL

### PR DESCRIPTION
**Pull request by honglei**
See https://github.com/tiangolo/sqlmodel/issues/532#issuecomment-1675860171

### CODE REVIEW

**Comment by mbsantiago** [comment]
Hi @honglei! Thanks for the PR. I've run the checks and they are failing. It seems to be some typing issues. I'm a bit busy atm so can't work this week, but I could help with some checks the next one.

File: sqlmodel/main.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
 from typing import (
     AbstractSet,
     Any,
+    Annotated,
```
Comment (mbsantiago) [inline sqlmodel/main.py]: Annotated was introduced in Python 3.9. Need to check python version and import it from typing_extensions if necessary

File: sqlmodel/main.py [hunk 2]
```diff
@@ -166,6 +170,18 @@ def Field(
     schema_extra: Optional[Dict[str, Any]] = None,
 ) -> Any:
     current_schema_extra = schema_extra or {}
+    if default is PydanticUndefined:
+        if type(sa_column) is types.FunctionType: #lambda
```
Comment (mbsantiago) [inline sqlmodel/main.py]: There seems to be an error with this line:

> Non-overlapping identity check (left operand type: "Union[Type[Column[Any]], Type[PydanticUndefinedType]]", right operand type: "Type[FunctionType]")

File: sqlmodel/main.py [hunk 3]
```diff
@@ -166,6 +170,18 @@ def Field(
...
     current_schema_extra = schema_extra or {}
+    if default is PydanticUndefined:
+        if type(sa_column) is types.FunctionType: #lambda 
+            sa_column_ = sa_column()
+        else:
+            sa_column_ = sa_column
+            
+        #assert type(sa_column) is Column:
+        if isinstance(sa_column_, Column) and sa_column_.server_default is not None:
+            default_value =  sa_column_.server_default.arg
```
Comment (mbsantiago) [inline sqlmodel/main.py]: > "FetchedValue" has no attribute "arg"

File: sqlmodel/main.py [hunk 4]
```diff
@@ -408,14 +424,27 @@ def __init__(
 def get_sqlalchemy_type(field: FieldInfo) -> Any:
     type_: Optional[type] = field.annotation
 
-    # Resolve Optional fields
-    if type_ is not None and get_origin(type_) is Union:
+    # Resolve Optional/Union fields
+    if type_ is not None and get_origin(type_) in (types.UnionType, Union):
```
Comment (mbsantiago) [inline sqlmodel/main.py]: Nice :)

File: sqlmodel/main.py [hunk 5]
```diff
@@ -408,14 +424,27 @@ def __init__(
...
+    # Resolve Annoted fields,
+    # like typing.Annotated[pydantic_core._pydantic_core.Url, 
+    # UrlConstraints(max_length=512, 
+    # allowed_schemes=['smb', 'ftp', 'file']) ]
+    if type_ is pydantic.AnyUrl:
+        meta =  field.metadata[0]
+        return AutoString(length=meta.max_length)
+    
+    if  get_origin(type_) is Annotated:
+        type2 = type_.__args__[0]
```
Comment (mbsantiago) [inline sqlmodel/main.py]: Mypy does not like this line:

>  Item "type" of "Optional[type]" has no attribute "__args__"

Maybe use https://docs.python.org/3/library/typing.html#typing.get_args

File: sqlmodel/main.py [hunk 6]
```diff
@@ -408,14 +424,27 @@ def __init__(
...
+        if type2 is pydantic.AnyUrl:
+            meta =  type_.__metadata__[0]
```
Comment (mbsantiago) [inline sqlmodel/main.py]: > Item "None" of "Optional[type]" has no attribute "__metadata__"

File: sqlmodel/main.py [hunk 7]
```diff
@@ -466,7 +495,7 @@ def get_sqlalchemy_type(field: FieldInfo) -> Any:
 
 def get_column_from_field(field: FieldInfo) -> Column:  # type: ignore
     sa_column = getattr(field, "sa_column", PydanticUndefined)
-    if isinstance(sa_column, Column):
+    if isinstance(sa_column, Column) or isinstance(sa_column, MappedColumn) :
         return sa_column
```
Comment (mbsantiago) [inline sqlmodel/main.py:524]: > Incompatible return value type (got "Union[Column[Any], MappedColumn[Any]]", expected "Column[Any]")

**Review by mbsantiago — CHANGES_REQUESTED** [review]
Many thanks for the PR! Please check this typing errors

File: sqlmodel/main.py (hunk 1 above)
Comment (honglei) [reply sqlmodel/main.py]: Ok, use typing_extensions to import `Annotated`.

File: sqlmodel/main.py (hunk 2 above)
Comment (honglei) [reply sqlmodel/main.py]: sa_column maybe a FunctionType which return a Column: https://github.com/sqlalchemy/sqlalchemy/issues/5669#issuecomment-715338925 
I'll change it.

File: sqlmodel/main.py (hunk 3 above)
Comment (honglei) [reply sqlmodel/main.py]: will use `DefaultClause`

File: sqlmodel/main.py (hunk 5 above)
Comment (honglei) [reply sqlmodel/main.py]: Thanks! [typing.get_args](https://docs.python.org/3/library/typing.html#typing.get_args) is much better.

File: sqlmodel/main.py (hunk 6 above)
Comment (honglei) [reply sqlmodel/main.py]: will use [typing.get_args](https://docs.python.org/3/library/typing.html#typing.get_args)

File: sqlmodel/main.py [hunk 8]
```diff
@@ -44,16 +44,16 @@
...
 
 from .sql.sqltypes import GUID, AutoString
 from .typing import SQLModelConfig
 
 if sys.version_info >= (3, 8):
-    from typing import get_args, get_origin
+    from typing import get_args, get_origin, Annotated
 else:
-    from typing_extensions import get_args, get_origin
+    from typing_extensions import get_args, get_origin, Annotated
```
Comment (mbsantiago) [inline sqlmodel/main.py]: Annotated is only available in 3.9 and above.

File: sqlmodel/main.py [hunk 9]
```diff
@@ -26,6 +25,7 @@
     TypeVar,
     Union,
     cast,
+    get_args
```
Comment (mbsantiago) [inline sqlmodel/main.py]: get_args gets imported afterwards

File: sqlmodel/main.py (hunk 3 above)
Comment (mbsantiago) [reply sqlmodel/main.py]: Great

File: sqlmodel/main.py [hunk 10]
```diff
@@ -158,14 +162,26 @@ def Field(
...
+        if type(sa_column) is types.FunctionType: #lambda 
+            sa_column_ = sa_column()
+        else:
+            sa_column_ = sa_column
+            
+        #server_default -> default
+        if isinstance(sa_column_, Column) and isinstance(sa_column_.server_default, DefaultClause):
+            default_value =  sa_column_.server_default.arg
+            if issubclass( type(sa_column_.type), Integer):
+                default = int(default_value)
```
Comment (mbsantiago) [inline sqlmodel/main.py:192]: > Argument 1 to "int" has incompatible type "Union[str, ClauseElement, TextClause]"; expected "Union[str, bytes, array[Any], mmap, _CData, PickleBuffer, SupportsInt, SupportsIndex, SupportsTrunc]"

**Review by mbsantiago — CHANGES_REQUESTED** [review]
Small fixes

File: sqlmodel/main.py (hunk 9 above)
Comment (honglei) [reply sqlmodel/main.py]: removed.

File: sqlmodel/main.py (hunk 10 above)
Comment (honglei) [reply sqlmodel/main.py:192]: limit to str type

**Comment by mbsantiago** [comment]
@honglei  Almost all typing errors are fixed. But there are some formatting errors now :/. Please consult the workflow logs to see the issues.

**Comment by honglei** [comment]
> @honglei Almost all typing errors are fixed. But there are some formatting errors now :/. Please consult the workflow logs to see the issues.

I use  ruff checked in Python3.7/debain10, all passed except code length.

**Pull request merged**